digdag-core/src/main/java/io/digdag/core/database/DatabaseMigrator.java

```diff
@@ -626,11 +626,117 @@ public void migrate(Handle handle)
+                    .addInt("queue_id", "")
+                    .addInt("priority", "not null")
+                    .addInt("retry_count", "not null default 0")
+                    .addLong("lock_expire_time", "")
+                    .addString("lock_agent_id", "")
+                    .build());
+
+            handle.update("insert into queued_task_locks" +
+                    " (id, site_id, queue_id, priority)" +
+                    " select id, site_id, NULL, priority" +
+                    " from queued_tasks");
+
+            // queues
+            handle.update("alter table queues" +
+                    " add column shared_site_id int");
+
+            // resource_types
+            handle.update("drop table resource_types");
+
+            // queued_tasks
+            handle.update("alter table queued_tasks" +
+                    " alter column task_id drop not null");
+            handle.update("alter table queued_tasks" +
+                    " alter column queue_id drop not null");
+            handle.update("alter table queued_tasks" +
+                    " drop column priority");
+            handle.update("alter table queued_tasks" +
+                    " drop column resource_type_id");
+            handle.update("create unique index queued_tasks_on_site_id_task_id on queued_tasks (site_id, task_id)");
+
+            if (isPostgres()) {
+                handle.update(
+                    "CREATE FUNCTION lock_shared_tasks(target_site_id int, target_site_max_concurrency bigint, limit_count int, lock_expire_seconds int, agent_id text) returns setof bigint as $$\n" +
+                    "BEGIN\n" +
+                    "  IF pg_try_advisory_xact_lock(23300, target_site_id) THEN\n" +
+                    "    RETURN QUERY\n" +
+                    "      with updated as (\n" +
+                    "        update queued_task_locks\n" +
+                    "        set lock_expire_time = cast(extract(epoch from statement_timestamp()) as bigint) + lock_expire_seconds,\n" +
+                    "            lock_agent_id = agent_id\n" +
+                    "        where id = any(\n" +
+                    "          select queued_task_locks.id\n" +
+                    "          from queued_task_locks\n" +
+                    "          where lock_expire_time is null\n" +
+                    "          and site_id = target_site_id\n" +
+                    "          and not exists (\n" +
+                    "            select * from (\n" +
+                    "              select queue_id, count(*) as count\n" +
+                    "              from queued_task_locks\n" +
+                    "              where lock_expire_time is not null\n" +
+                    "                and site_id = target_site_id\n" +
+                    "              group by queue_id\n" +
+                    "            ) runnings\n" +
+                    "            join queues on queues.id = runnings.queue_id\n" +
```

Comment:
Do you mean like this?:

```
...
and not exists (
  select queue_id, count(*) as count
  from queued_task_locks
  join queues on queues.id = queued_task_locks.queue_id
  where lock_expire_time is not null
    and site_id = 1
    and queued_task_locks.queue_id = queued_task_locks.queue_id
  group by queue_id
  having count(*) >= queues.max_concurrency
)
...
```

This query is invalid because HAVING requires GROUP BY keys or aggregation functions but queues.max_concurrency is none of them:

```
ERROR:  column "queues.max_concurrency" must appear in the GROUP BY clause or be used in an aggregate function
LINE 14:   having count(*) >= queues.max_concurrency
```